### PR TITLE
Umount injection

### PR DESCRIPTION
**Pull request by LizaTretyakova-zz**

### CODE REVIEW

File: src/lxc/lxccontainer.c [hunk 1]
```diff
@@ -4840,7 +4840,66 @@ static int do_lxcapi_mount(struct lxc_container *c,
...
+	if (pid == 0) {
+		init_pid = do_lxcapi_init_pid(c);
+		if (init_pid < 0) {
+			ERROR("Failed to obtain container's init pid");
+			_exit(EXIT_FAILURE);
+		}
+	
+		/* Enter the container namespaces */
+		if (am_guest_unpriv() && !lxc_list_empty(&c->lxc_conf->id_map)) {
+			/* if fully unprivileged */
```
Comment (brauner) [inline src/lxc/lxccontainer.c]: I think you can remove the `am_guest_unpriv()` and simply rely on the `lxc_list_empty` check. :)

Comment (brauner) [reply src/lxc/lxccontainer.c]: You can also remove the comment.

File: src/lxc/lxccontainer.c [hunk 2]
```diff
@@ -4840,7 +4840,66 @@ static int do_lxcapi_mount(struct lxc_container *c,
...
+				ERROR("Failed to enter user namespace");
+				_exit(EXIT_FAILURE);
+			}
+		}
+		if (!switch_to_ns(init_pid, "mnt")) {
+			ERROR("Failed to enter mount namespace");
+			_exit(EXIT_FAILURE);
+		}
+
+		/* Do the mount */
```
Comment (brauner) [inline src/lxc/lxccontainer.c]: s/Do the mount/do the umount/

File: src/lxc/lxccontainer.c [hunk 3]
```diff
@@ -4840,7 +4840,66 @@ static int do_lxcapi_mount(struct lxc_container *c,
...
+		ret = umount2(target, mountflags);
+		if (ret < 0) {
+			/* TODO: deal with MNT_EXPIRE? */
```
Comment (brauner) [inline src/lxc/lxccontainer.c]: You can remove that comment too. :)

File: src/lxc/lxccontainer.h [hunk 4]
```diff
@@ -859,6 +859,12 @@ struct lxc_container {
 				 const char *source, const char *target,
 				 const char *filesystemtype, unsigned long mountflags,
 				 const void *data, struct lxc_mount *mnt);
+
+	/*!
+	 * \brief Unmount the container's path `target`.
+	 */
+	int (*umount2)(struct lxc_container *c, const char *target,
```
Comment (brauner) [inline src/lxc/lxccontainer.h]: I think we should just call it `umount()` instead of `umount2()`. :)

**Review by brauner — CHANGES_REQUESTED** [review]
Cool!

File: src/lxc/lxccontainer.c (hunk 1 above)
Comment (LizaTretyakova-zz) [reply src/lxc/lxccontainer.c]: Why can we do it here? I mean, remove the `am_guest_unpriv()` call O.o

File: src/lxc/lxccontainer.c (hunk 2 above)
Comment (LizaTretyakova-zz) [reply src/lxc/lxccontainer.c]: Oops ^^

File: src/lxc/lxccontainer.c (hunk 3 above)
Comment (LizaTretyakova-zz) [reply src/lxc/lxccontainer.c]: Btw, do we actually want to deal with it? :) I mean, the error might be EAGAIN and stuff...

File: src/lxc/lxccontainer.c (hunk 1 above)
Comment (brauner) [reply src/lxc/lxccontainer.c]: Oh, you're interested in the fully-unprivileged use-case. Hm, I guess we can leave it in then. :)

File: src/lxc/lxccontainer.c (hunk 3 above)
Comment (brauner) [reply src/lxc/lxccontainer.c]: Ok, what we should do is:

If MNT_EXPIRE is specified and we get EAGAIN, we should make sure that
`errno` is preserved and that we return `-EAGAIN` as return value to the caller.
This way callers can decide whether they want to call `umount()` again or not.

In general, weshould make sure that on error we always return the negative
`errno` value to the caller.

File: src/lxc/lxccontainer.c (hunk 1 above)
Comment (brauner) [reply src/lxc/lxccontainer.c]: Right, so we discussed this last week. We need to remove the `am_guest_unpriv()` check. :)

**Comment by LizaTretyakova-zz** [comment]
Moved to the https://github.com/lxc/lxc/pull/2300

**Pull request closed without merging**